File: 1354-construct-target-array-with-multiple-sums/1354-construct-target-array-with-multiple-sums.cpp
#define ll long long
class Solution {
public:
    bool isPossible(vector<int>& target) {
        
        priority_queue<ll> pq;
        ll total_sum=0;
        
        for(auto it: target){
            pq.push(it);
            total_sum += it;   
        }
        
        while(pq.size()>0){
            
            ll top = pq.top(); pq.pop();
            ll remaining_sum = total_sum - top;
            
            if(top==1 || remaining_sum ==1) return true;
            
            if(remaining_sum==0 || top < remaining_sum) return false;
            
            int val = (int)(top%remaining_sum);
            
            if(val==0) return false;
            
            pq.push(val);
            total_sum = remaining_sum + val;
        }
        
        return true;
    }
};

File: 1354-construct-target-array-with-multiple-sums/1354-construct-target-array-with-multiple-sums.py
class Solution:
    def isPossible(self, target: List[int]) -> bool:
        
        totalSum=0
        heap = []
        heapify(heap)
        
        for el in target:
            heappush(heap, -1*el)
            totalSum += el
        
        while(len(heap) > 0):
            
            maxEl = -1 * heappop(heap)
            # print(maxEl)
            remTotal = totalSum - maxEl
            
            if maxEl == 1 or remTotal == 1:
                return True
            
            if remTotal == 0 or maxEl < remTotal:
                return False
            
            updateMax = maxEl % remTotal
            
            if updateMax == 0:
                return False
            
            heappush(heap, -1*updateMax)
            
            totalSum = remTotal + updateMax
        
        return True